contracts/utils/structs/MaxHeap.sol

```diff
@@ -0,0 +1,162 @@
+ *
+ * A max heap is a complete binary tree where each node has a value greater than or equal to its children.
+ * The root node contains the maximum value in the heap.
+ *
+ * This library provides functions to insert, update, and remove elements from the max heap, as well as to
+ * retrieve the maximum element (peek) and check the validity of the heap.
+ *
+ * The max heap has the following properties:
+ *
+ * - Insertion: O(log n)
+ * - Deletion of maximum element: O(log n)
+ * - Retrieval of maximum element (peek): O(1)
+ * - Update of an element: O(log n)
+ *
+ * The max heap is implemented using two mappings:
+ * - `tree`: Maps the position in the heap to the item ID.
+ * - `items`: Maps the item ID to its corresponding `Node` struct, which contains the value and heap index.
+ *
+ * Example usage:
+ *
+ * ```solidity
+ * contract Example {
+ *     using MaxHeap for MaxHeap.MaxHeap;
+ *
+ *     MaxHeap.MaxHeap private heap;
+ *
+ *     function addItem(uint256 itemId, uint256 value) public {
+ *         heap.insert(itemId, value);
+ *     }
+ *
+ *     function removeMax() public returns (uint256, uint256) {
+ *         return heap.pop();
+ *     }
+ *
+ *     function getMax() public view returns (uint256, uint256) {
+ *         return heap.peek();
+ *     }
+ * }
+ * ```
+ */
+library MaxHeap {
+    /**
+     * @dev The position doesn't have a _parent as it's the root.
+     */
+    error InvalidPositionZero();
+
+    struct Node {
+        uint256 value;
+        uint256 heapIndex;
+    }
+
+    struct MaxHeap {
+        mapping(uint256 => uint256) tree;
+        mapping(uint256 => Node) items;
+        uint256 size;
+    }
+
+    function _parent(uint256 pos) private pure returns (uint256) {
+        if (pos == 0) revert InvalidPositionZero();
+        return (pos - 1) / 2;
+    }
+
+    function _swap(MaxHeap storage heap, uint256 fpos, uint256 spos) private {
+        (heap.tree[fpos], heap.tree[spos]) = (heap.tree[spos], heap.tree[fpos]);
+        (heap.items[heap.tree[fpos]].heapIndex, heap.items[heap.tree[spos]].heapIndex) = (fpos, spos);
+    }
+
+    function heapify(MaxHeap storage heap, uint256 pos) internal {
+        if (pos >= (heap.size / 2) && pos <= heap.size) return;
+
+        uint256 left = 2 * pos + 1;
+        uint256 right = left + 1;
+
+        uint256 leftValue = left < heap.size ? heap.items[heap.tree[left]].value : 0;
+        uint256 rightValue = right < heap.size ? heap.items[heap.tree[right]].value : 0;
+        uint256 posValue = heap.items[heap.tree[pos]].value;
+
+        if (posValue < leftValue || posValue < rightValue) {
+            if (leftValue > rightValue) {
+                _swap(heap, pos, left);
+                heapify(heap, left);
+            } else {
+                _swap(heap, pos, right);
+                heapify(heap, right);
+            }
+        }
+    }
+
+    function insert(MaxHeap storage heap, uint256 itemId, uint256 value) internal {
```

Comment:
My understand is that itemId must be unique, so that all items are clearly identified ... but is then useless to the user. Maybe it should be automatically generated ... so that the user doesn't have to worry about it, and can just push a value.